### PR TITLE
Replaced TreeHashVisitor with an enhanced JexlStringBuildingVisitor t…

### PR DESCRIPTION
**Pull request by jwomeara**
…o fix a NPE and updated usages throughout.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlStringBuildingVisitor.java [hunk 1]
```diff
@@ -154,14 +196,14 @@ public Object visit(ASTAndNode node, Object data) {
             sb.append("(");
         }
         
-        int lastsize = sb.length();
+        Collection<String> childStrings = (sortDedupeChildren) ? new TreeSet<>() : new ArrayList<>();
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlStringBuildingVisitor.java]: The ArrayList could take the number of child nodes as an argument.

Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlStringBuildingVisitor.java]: Sure.  See latest commit.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java [hunk 2]
```diff
@@ -692,4 +691,8 @@ private boolean isInfinite() {
     public void setNode(JexlNode currNode) {
         myNode = currNode;
     }
+    
+    private String nodeToKey(JexlNode node) {
+        return JexlStringBuildingVisitor.buildQueryWithoutParse(TreeFlatteningRebuildingVisitor.flatten(node), true);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java]: we call this a lot.  Is it possible to flatten the tree once?

Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java]: Per our conversation, I verified that in all of our test cases the node passed in to nodeToKey is already flattened.  We should be able to safely remove the flatten call from this method.

File: warehouse/query-core/src/main/java/datawave/query/jexl/DatawaveInterpreter.java [hunk 3]
```diff
@@ -83,9 +82,9 @@ public static boolean isMatched(Object scriptExecuteResult) {
     
     @Override
     public Object visit(ASTFunctionNode node, Object data) {
-        TreeHashNode hash = TreeHashVisitor.getNodeHash(node);
+        String nodeString = JexlStringBuildingVisitor.buildQuery(node);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/DatawaveInterpreter.java]: buildQueryWithoutParse?

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java [hunk 4]
```diff
@@ -692,4 +691,9 @@ private boolean isInfinite() {
     public void setNode(JexlNode currNode) {
         myNode = currNode;
     }
+    
+    private String nodeToKey(JexlNode node) {
+        // Note: This method assumes that the node passed in is already flattened.
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java]: The cost of not being flattened is....

**Pull request merged**